scriptworker/cot/verify.py

```diff
@@ -1228,41 +1082,101 @@ def verify_decision_command(decision_link, mach_commands=DECISION_MACH_COMMANDS)
-    """Get the appropriate json-e template.
+# get_jsone_context_and_template {{{1
+async def get_in_tree_template(link):
+    """Get the in-tree json-e template for a given link.
+
+    By convention, this template is SOURCE_REPO/.taskcluster.yml.
 
     Args:
+        link (LinkOfTrust): the parent link to get the source url from.
+
+    Raises:
+        CoTError: on non-yaml `source_url`
+        KeyError: on non-well-formed source template
+
+    Returns:
+        dict: the first task in the template.
+
+    """
+    context = link.context
+    source_url = get_source_url(link)
+    if not source_url.endswith(('.yml', '.yaml')):
+        raise CoTError("{} source url {} doesn't end in .yml or .yaml!".format(
+            link.name, source_url
+        ))
+    tmpl = await load_json_or_yaml_from_url(
+        context, source_url, os.path.join(
+            context.config["work_dir"], "{}_taskcluster.yml".format(link.name)
+        )
+    )
+    return tmpl['tasks'][0]
+
+
+async def get_action_context_and_template(chain, parent_link, decision_link):
+    """Get the appropriate json-e context and template for an action task.
+
+    Args:
+        chain (ChainOfTrust): the chain of trust.
+        parent_link (LinkOfTrust): the parent link to test.
+        decision_link (LinkOfTrust): the parent link's decision task link.
+        tasks_for (str): the reason the parent link was created (cron,
+            hg-push, action)
+
+    Returns:
+        (dict, dict): the json-e context and template.
+
+    """
+    actions_path = decision_link.get_artifact_full_path('public/actions.json')
+    all_actions = load_json_or_yaml(actions_path, is_path=True)['actions']
+    action_name = get_action_name(parent_link.task)
+    action_defn = [d for d in all_actions if d['name'] == action_name][0]
+    jsone_context = await populate_jsone_context(chain, parent_link, decision_link, "action")
+    if 'task' in action_defn and chain.context.config['min_cot_version'] <= 2:
+        tmpl = action_defn['task']
+    else:
+        tmpl = await get_in_tree_template(decision_link)
+        for k in ('action', 'push', 'repository'):
+            jsone_context[k] = deepcopy(action_defn['hookPayload']['decision'][k])
+        jsone_context['action']['repo_scope'] = get_repo_scope(parent_link.task)
+
+    return jsone_context, tmpl
+
+
+async def get_jsone_context_and_template(chain, parent_link, decision_link, tasks_for):
+    """Get the appropriate json-e context and template for any parent task.
+
+    Args:
+        chain (ChainOfTrust): the chain of trust.
         parent_link (LinkOfTrust): the parent link to test.
         decision_link (LinkOfTrust): the parent link's decision task link.
         tasks_for (str): the reason the parent link was created (cron,
             hg-push, action)
 
     Returns:
-        dict: the json-e template.
+        (dict, dict): the json-e context and template.
 
     """
     if tasks_for == 'action':
```

Comment:
I'm wondering, where do we safeguard against `tasks_for` not being in `('action', 'decision')`?

Reply:
[`get_and_check_tasks_for`](https://github.com/mozilla-releng/scriptworker/blob/master/scriptworker/task.py#L216-L240) throws if `tasks_for` isn't in [`KNOWN_TASKS_FOR`](https://github.com/mozilla-releng/scriptworker/blob/master/scriptworker/task.py#L28) .